Discovery/Discovery/Presentation/WebPrograms/ProgramWebviewViewModel.swift

```diff
@@ -0,0 +1,239 @@
+        var externalLink = false
+        
+        if let queryParameters = request.url?.queryParameters,
+            let externalLinkValue = queryParameters["external_link"] as? String,
+           externalLinkValue.caseInsensitiveCompare("true") == .orderedSame {
+            externalLink = true
+        }
+        
+        if let url = request.url, outsideLink || capturedLink || externalLink, UIApplication.shared.canOpenURL(url) {
+            DispatchQueue.main.async { [weak self] in
+                self?.router.presentAlert(
+                    alertTitle: DiscoveryLocalization.Alert.leavingAppTitle,
+                    alertMessage: DiscoveryLocalization.Alert.leavingAppMessage,
+                    positiveAction: CoreLocalization.Webview.Alert.continue,
+                    onCloseTapped: {
+                        self?.router.dismiss(animated: true)
+                    }, okTapped: {
+                        UIApplication.shared.open(url, options: [:])
+                    }, type: .default(positiveAction: CoreLocalization.Webview.Alert.continue)
+                )
+            }
+            return true
+        }
+        
+        return false
+    }
+    
+    private func urlAction(from url: URL) -> WebviewActions? {
+        guard isValidAppURLScheme(url),
+                let url = WebviewActions(rawValue: url.appURLHost) else { return nil }
+        return url
+    }
+    
+    @MainActor
+    private  func handleNavigation(url: URL, urlAction: WebviewActions) -> Bool {
+        switch urlAction {
+        case .courseEnrollment:
+            if let urlData = parse(url: url), let courseID = urlData.courseId {
+                Task {
+                    await enrollTo(courseID: courseID)
+                }
+            }
+        case .courseDetail:
+            guard let pathID = detailPathID(from: url) else { return false }
+            router.showWebDiscoveryDetails(
+                pathID: pathID,
+                discoveryType: .courseDetail(pathID),
+                sourceScreen: .default
+            )
+        case .enrolledCourseDetail:
+            if let urlData = parse(url: url), let courseID = urlData.courseId {
+                showProgress = true
+                Task {
+                    do {
+                        courseDetails = try await getCourseDetail(courseID: courseID)
+                        showCourseDetails()
+                        courseDetails = nil
+                        showProgress = false
+                    } catch let error {
+                        showProgress = false
+                        if error.isInternetError || error is NoCachedDataError {
+                            errorMessage = CoreLocalization.Error.slowOrNoInternetConnection
+                        } else {
+                            errorMessage = CoreLocalization.Error.unknownError
+                        }
+                    }
+                }
+                
+            }
```

Comment:
Is it possible to convert the `handleNavigation` function to `async` and remove the creation of Tasks in it?
It will be difficult to test with a Task inside it.

Reply:
Done

Reply:
Thank you!